### PR TITLE
[SPARK-25981][R] Enables Arrow optimization from R DataFrame to Spark DataFrame

### PR DESCRIPTION
**Pull request by HyukjinKwon**
## What changes were proposed in this pull request?

This PR targets to support Arrow optimization for conversion from R DataFrame to Spark DataFrame.
Like PySpark side, it falls back to non-optimization code path when it's unable to use Arrow optimization.

This can be tested as below:

```bash
$ ./bin/sparkR --conf spark.sql.execution.arrow.enabled=true
```

```r
collect(createDataFrame(mtcars))
```

### Requirements
  - R 3.5.x 
  - Arrow package 0.12+
    ```bash
    Rscript -e 'remotes::install_github("apache/arrow@apache-arrow-0.12.0", subdir = "r")'
    ```

**Note:** currently, Arrow R package is not in CRAN. Please take a look at ARROW-3204.
**Note:** currently, Arrow R package seems not supporting Windows. Please take a look at ARROW-3204.

### Benchmarks

**Shall**

```bash
sync && sudo purge
./bin/sparkR --conf spark.sql.execution.arrow.enabled=false
```

```bash
sync && sudo purge
./bin/sparkR --conf spark.sql.execution.arrow.enabled=true
```

**R code**

```r
createDataFrame(mtcars) # Initializes
rdf <- read.csv("500000.csv")

test <- function() {
  options(digits.secs = 6) # milliseconds
  start.time <- Sys.time()
  createDataFrame(rdf)
  end.time <- Sys.time()
  time.taken <- end.time - start.time
  print(time.taken)
}

test()
```

**Data (350 MB):**

```r
object.size(read.csv("500000.csv"))
350379504 bytes
```

"500000 Records"  http://eforexcel.com/wp/downloads-16-sample-csv-files-data-sets-for-testing/

**Results**

```
Time difference of 29.9468 secs
```

```
Time difference of 3.222129 secs
```

The performance improvement was around **950%**.
Actually, this PR improves around **1200%**+ because this PR includes a small optimization about regular R DataFrame -> Spark DatFrame. See https://github.com/apache/spark/pull/22954#discussion_r231847272

### Limitations:

For now, Arrow optimization with R does not support when the data is `raw`, and when user explicitly gives float type in the schema. They produce corrupt values.
In this case, we decide to fall back to non-optimization code path.

## How was this patch tested?

Small test was added.

I manually forced to set this optimization `true` for _all_ R tests and they were _all_ passed (with few of fallback warnings).

**TODOs:**
- [x] Draft codes
- [x] make the tests passed
- [x] make the CRAN check pass
- [x] Performance measurement
- [x] Supportability investigation (for instance types)
- [x] Wait for Arrow 0.12.0 release
- [x] Fix and match it to Arrow 0.12.0

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
Let me leave a cc @felixcheung, @BryanCutler FYI.

**Comment by HyukjinKwon** [comment]
Adding @yanghaogn as well fyi

**Comment by HyukjinKwon** [comment]
So far, the regressions tests are passed and newly added test for R optimization is verified locally. Let me fix CRAN test and some nits.

File: R/pkg/R/SQLContext.R
Comment (felixcheung) [reply R/pkg/R/SQLContext.R]: btw, is it worthwhile to check the arrow package version?

File: R/pkg/R/SQLContext.R
Comment (felixcheung) [inline R/pkg/R/SQLContext.R]: I think you can use sparkR.conf

**Review by felixcheung — COMMENTED** [review]
so this is super cool - my biggest concern is we changed to not write out to file for not respecting the encryption requirement, and this adds back the temp file

File: R/pkg/R/SQLContext.R
Comment (felixcheung) [inline R/pkg/R/SQLContext.R]: might need to give it a dir prefix to use - the tempfile default is not CRAN compliant and possibly some ACL issue

**Comment by HyukjinKwon** [comment]
For encryption stuff, I will try to handle that as well (maybe as a followup(?)) so that we support it even when that's enabled.

**Comment by HyukjinKwon** [comment]
@felixcheung! performance improvement was **955%** ! I described the benchmark I took in PR description.

File: R/pkg/R/SQLContext.R
Comment (HyukjinKwon) [inline R/pkg/R/SQLContext.R]: Note that this PR optimizes the original code path as well here - when the input is local R DataFrame, here we avoid `firstRDD` operation.

In the master branch, the benchmark shows:

```
Exception in thread "dispatcher-event-loop-6" java.lang.OutOfMemoryError: Java heap space
	at java.util.Arrays.copyOf(Arrays.java:3236)
	at java.io.ByteArrayOutputStream.grow(ByteArrayOutputStream.java:118)
	at java.io.ByteArrayOutputStream.ensureCapacity(ByteArrayOutputStream.java:93)
	at java.io.ByteArrayOutputStream.write(ByteArrayOutputStream.java:153)
	at org.apache.spark.util.ByteBufferOutputStream.write(ByteBufferOutputStream.scala:41)
	at java.io.ObjectOutputStream$BlockDataOutputStream.drain(ObjectOutputStream.java:1877)
```

So, technically this PR improves

If I try this with `100000.csv` (79MB) record, it takes longer to cut it short:

**Current master**:

```
Time difference of 8.502607 secs
```

**With this PR, but without Arrow**

```
Time difference of 5.143395 secs
```

**With this PR, but with Arrow**


```
Time difference of 0.6981369 secs
```

So, technically this PR improves more **1200%**

**Comment by HyukjinKwon** [comment]
adding @falaki and @mengxr as well.

File: R/pkg/R/SQLContext.R
Comment (HyukjinKwon) [inline R/pkg/R/SQLContext.R]: These workarounds will be removed when Arrow 0.12.0 is released. I did it to make CARN passed.

**Comment by HyukjinKwon** [comment]
I have finished most of todos except waiting for R API of Arrow 0.12.0 and fixing some changes accordingly.

**Comment by HyukjinKwon** [comment]
retest this please

File: R/pkg/R/SQLContext.R
Comment (HyukjinKwon) [inline R/pkg/R/SQLContext.R]: It's actually a bit odd that I need to manually require this package. Otherwise, it complains, for instance, here https://github.com/apache/arrow/blob/d3ec69069649013229366ebe01e22f389597dc19/r/R/on_exit.R#L20

File: R/pkg/R/SQLContext.R
Comment (felixcheung) [reply R/pkg/R/SQLContext.R]: I <3 4 digits!

File: R/pkg/R/SQLContext.R
Comment (felixcheung) [reply R/pkg/R/SQLContext.R]: ok

File: R/pkg/R/SQLContext.R
Comment (felixcheung) [reply R/pkg/R/SQLContext.R]: require1 sounds a bit like a hack though...

File: sql/core/src/main/scala/org/apache/spark/sql/api/r/SQLUtils.scala
Comment (felixcheung) [inline sql/core/src/main/scala/org/apache/spark/sql/api/r/SQLUtils.scala]: what's the advantage of adding this wrapper here - I've thinking to eliminate most of these if possible - and just use callJMethod on `ArrowConverters` say?

File: R/pkg/R/SQLContext.R
Comment (felixcheung) [inline R/pkg/R/SQLContext.R]: should this go by default with default parallelism ?

File: R/pkg/R/SQLContext.R
Comment (felixcheung) [inline R/pkg/R/SQLContext.R]: `1 : ceiling`? `1 : nrow`?

Comment (felixcheung) [reply R/pkg/R/SQLContext.R]: how is this slices computed? is it similar to https://github.com/apache/spark/blob/c3b4a94a91d66c172cf332321d3a78dba29ef8f0/R/pkg/R/context.R#L152

File: R/pkg/R/SQLContext.R
Comment (felixcheung) [inline R/pkg/R/SQLContext.R]: so is this an API missing in Arrow?

Comment (felixcheung) [reply R/pkg/R/SQLContext.R]: actually, I think if you use withr here it will call close_on_exit for you? (but when?)

File: R/pkg/R/SQLContext.R
Comment (felixcheung) [inline R/pkg/R/SQLContext.R]: is it always in the conf - I think you can also pass in a default value to sparkR.conf

File: R/pkg/R/SQLContext.R
Comment (felixcheung) [inline R/pkg/R/SQLContext.R]: can you check -  I think `is` `is.x` doesn't something do the right thing when

head(df, 1) and one of the field is `NA`

File: R/pkg/R/SQLContext.R
Comment (felixcheung) [inline R/pkg/R/SQLContext.R]: ? https://stat.ethz.ch/R-manual/R-devel/library/base/html/warning.html

File: R/pkg/R/SQLContext.R
Comment (felixcheung) [inline R/pkg/R/SQLContext.R]: unlink?

File: R/pkg/R/SQLContext.R
Comment (felixcheung) [inline R/pkg/R/SQLContext.R]: refactor this into a method?

File: R/pkg/R/SQLContext.R
Comment (HyukjinKwon) [reply R/pkg/R/SQLContext.R]: ooops

File: sql/core/src/main/scala/org/apache/spark/sql/api/r/SQLUtils.scala
Comment (BryanCutler) [inline sql/core/src/main/scala/org/apache/spark/sql/api/r/SQLUtils.scala]: nit: a `RDD` -> an `RDD`

File: R/pkg/R/SQLContext.R
Comment (BryanCutler) [inline R/pkg/R/SQLContext.R]: Any idea what's going on with the `FloatType`? Is it a problem on the arrow side?

**Comment by BryanCutler** [comment]
I don't know R well enough to review that code, but the results look awesome! Nice work @HyukjinKwon!!

**Comment by HyukjinKwon** [comment]
Hey guys thanks for reviewing! Will address them soon.

File: R/pkg/R/SQLContext.R
Comment (kiszk) [inline R/pkg/R/SQLContext.R]: nit: `arrow` -> `Arrow`

File: R/pkg/R/SQLContext.R
Comment (HyukjinKwon) [reply R/pkg/R/SQLContext.R]: Yup .. it is .. looks we shouldn't have this error from a cursory look in R API of Arrow. Maybe this can be gone when I use official R Arrow release version. Let me check it later.

File: sql/core/src/main/scala/org/apache/spark/sql/api/r/SQLUtils.scala
Comment (HyukjinKwon) [reply sql/core/src/main/scala/org/apache/spark/sql/api/r/SQLUtils.scala]: Hmhmhm .. yea. What I was trying to do is to add SQL related codes called in R from JVM, into here when they are not official APIs in order to avoid, we change the internal APIs within Scala, and it causes R test failure. I was trying to do the similar things within PySpark side.

File: R/pkg/R/SQLContext.R
Comment (HyukjinKwon) [reply R/pkg/R/SQLContext.R]: We should; however, it follows the original code path's behaviour. I matched it as the same so that we can compare the performances in the same conditions. If you don't mind, I will fix both in a separate PR.

File: R/pkg/R/SQLContext.R
Comment (HyukjinKwon) [reply R/pkg/R/SQLContext.R]: This resembles PySpark side logic:

https://github.com/apache/spark/blob/d367bdcf521f564d2d7066257200be26b27ea926/python/pyspark/sql/session.py#L554-L556

Let me check the difference between them

File: R/pkg/R/SQLContext.R
Comment (HyukjinKwon) [reply R/pkg/R/SQLContext.R]: Hm, possibly yea. Let me try it. In this way, we could get rid of `require`.

File: R/pkg/R/SQLContext.R
Comment (HyukjinKwon) [reply R/pkg/R/SQLContext.R]: Yea,I checked that it always has the default value. I initially left the default value but took it out after double checking.

File: R/pkg/R/SQLContext.R
Comment (HyukjinKwon) [reply R/pkg/R/SQLContext.R]: Yup, let me try.

File: R/pkg/R/SQLContext.R
Comment (HyukjinKwon) [reply R/pkg/R/SQLContext.R]: I suspect that it happens when `numeric` (which is like `1.0`) is casted into float type. I think it's related with casting behaviour. Let me take a look and file a JIRA there in Arrow side but if you don't mind I will focus on matching exact type cases for now ...

File: R/pkg/R/SQLContext.R
Comment (HyukjinKwon) [reply R/pkg/R/SQLContext.R]: I believe either way is fine.

File: R/pkg/R/SQLContext.R
Comment (HyukjinKwon) [reply R/pkg/R/SQLContext.R]: Looks okay in this case specifically:

```r
> any(sapply(head(data.frame(list(list(a=NA))), 1), is.raw))
[1] FALSE
> any(sapply(head(data.frame(list(list(a=NA))), 1), function(x) is(x, "POSIXct")))
[1] FALSE
> any(sapply(head(data.frame(list(list(a=1))), 1), is.raw))
[1] FALSE
> any(sapply(head(data.frame(list(list(a="a"))), 1), function(x) is(x, "POSIXct")))
[1] FALSE
> any(sapply(head(data.frame(list(list(a=raw(1)))), 1), is.raw))
[1] TRUE
> any(sapply(head(data.frame(list(list(a=as.POSIXct("2000-01-01")))), 1), function(x) is(x, "POSIXct")))
[1] TRUE
```

File: R/pkg/R/SQLContext.R
Comment (HyukjinKwon) [reply R/pkg/R/SQLContext.R]: Let me try to reuse the R side slicing logic.

File: sql/core/src/main/scala/org/apache/spark/sql/api/r/SQLUtils.scala
Comment (felixcheung) [reply sql/core/src/main/scala/org/apache/spark/sql/api/r/SQLUtils.scala]: hmm, I see your point... but there could be hundreds of these wrapper we need add if we set as a practice, I'm guessing. 

a few problems with these wrappers I see:
1. they are extra work to add or maintain
2. many are very simple, not much value add
3. many get abandoned over the years - they are not called and not removed

but I kinda see your way, let's keep this one and review any new one in the future.

File: R/pkg/R/SQLContext.R
Comment (felixcheung) [reply R/pkg/R/SQLContext.R]: ah, any idea that was done that way in python? this seems to be different from sc.paralleize which is what https://github.com/apache/spark/blob/c3b4a94a91d66c172cf332321d3a78dba29ef8f0/R/pkg/R/context.R#L152 is done

File: R/pkg/R/SQLContext.R
Comment (felixcheung) [reply R/pkg/R/SQLContext.R]: ok

File: R/pkg/R/SQLContext.R
Comment (felixcheung) [reply R/pkg/R/SQLContext.R]: LG, I tested a few cases too

File: R/pkg/R/SQLContext.R
Comment (felixcheung) [reply R/pkg/R/SQLContext.R]: yes, just more consistent. I also don't know for sure why all other instances are calling unlink

File: R/pkg/R/SQLContext.R
Comment (felixcheung) [reply R/pkg/R/SQLContext.R]: maybe out of bit range? 53 bit https://stat.ethz.ch/R-manual/R-patched/library/base/html/double.html

File: R/pkg/R/SQLContext.R
Comment (felixcheung) [reply R/pkg/R/SQLContext.R]: that will be good. circumventing CRAN check for method name is... problematic..
(there are other more hacky way too, but ..)

File: R/pkg/R/SQLContext.R
Comment (HyukjinKwon) [reply R/pkg/R/SQLContext.R]: Not sure. I think the intention is the same. Let me stick to R's one for now.

**Comment by HyukjinKwon** [comment]
Let me hide some comments that are addressed (it looks messy). Please make unhide if I mistakenly hide some comments that are not addressed yet.

**Comment by HyukjinKwon** [comment]
Hm .. the CRAN passed in my local. Let me workaround for now.

File: R/pkg/tests/fulltests/test_sparkSQL.R
Comment (felixcheung) [inline R/pkg/tests/fulltests/test_sparkSQL.R]: does this fail? or just a way to inject a finally?

File: R/pkg/tests/fulltests/test_sparkSQL.R
Comment (felixcheung) [inline R/pkg/tests/fulltests/test_sparkSQL.R]: are we going to ask shane to install arrow/withr on the Jenkins machines?

File: R/pkg/R/SQLContext.R
Comment (felixcheung) [inline R/pkg/R/SQLContext.R]: future: consolidate the default here and inside makeSplits

File: R/pkg/R/SQLContext.R
Comment (felixcheung) [inline R/pkg/R/SQLContext.R]: nit: just `FALSE` is good

**Review by felixcheung — APPROVED** [review]
LGTM. a few workarounds are a bit nasty (withr, CRAN, close_on_exit) hopefully we could make them go away later.

about encryption, I think we should do this similar to before - ie. if encryption then do not use arrow/file. that should be enough

File: R/pkg/tests/fulltests/test_sparkSQL.R
Comment (HyukjinKwon) [reply R/pkg/tests/fulltests/test_sparkSQL.R]: Just to inject the finally .. :-) ..

File: R/pkg/tests/fulltests/test_sparkSQL.R
Comment (HyukjinKwon) [reply R/pkg/tests/fulltests/test_sparkSQL.R]: Maybe we should hold it for now .. because I realised R API for Arrow requires R 3.5.x and Jenkins's one is 3.1.x if I remember this correctly. Ideally, we could probably do that via AppVeyor if everything goes fine.

**Comment by HyukjinKwon** [comment]
Yea .. I will make the followup works right away after this one get merged. Thanks @felixcheung. Let me address the rest of comments, and wait for Arrow release.

@BryanCutler BTW, do you know the rough expected timing for Arrow 0.12.0 release?

File: R/pkg/R/SQLContext.R
Comment (viirya) [inline R/pkg/R/SQLContext.R]: Should we move `writeToTempFileInArrow` call into next `tryCatch`?

File: R/pkg/R/SQLContext.R
Comment (viirya) [inline R/pkg/R/SQLContext.R]: When `shouldUseArrow` is true, I think it means we already done using arrow? Maybe just `useArrow`?

File: sql/core/src/main/scala/org/apache/spark/sql/api/r/SQLUtils.scala
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/api/r/SQLUtils.scala]: Is this going to read a file in Arrow stream format?

File: R/pkg/R/SQLContext.R
Comment (HyukjinKwon) [reply R/pkg/R/SQLContext.R]: Yup, correct. Let me address other comments as well.

File: R/pkg/R/SQLContext.R
Comment (HyukjinKwon) [reply R/pkg/R/SQLContext.R]: I think it's a bug because it always produces a corrupt value when I try to use `number` as explicit float types.

**Review by felixcheung — APPROVED** [review]

**Comment by BryanCutler** [comment]
> @BryanCutler BTW, do you know the rough expected timing for Arrow 0.12.0 release?

I think we should be starting the release process soon, so maybe in a week or two.

**Comment by felixcheung** [comment]
FYI the release tracker (quite useful!) https://cwiki.apache.org/confluence/display/ARROW/Arrow+0.12.0+Release

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala
Comment (kiszk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: super nit if you will need additional commit: `... Pandas DataFrame," +` -> `... Pandas DataFrame, " +`

Comment (HyukjinKwon) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: Oh yeah. I need to push additional commit when it's released. Will fix it together later.

**Comment by felixcheung** [comment]
@HyukjinKwon it's released!

**Comment by felixcheung** [comment]
btwm @HyukjinKwon would you be interested in a post about this, similar to ARROW-4262?

**Comment by HyukjinKwon** [comment]
Sure!

**Comment by felixcheung** [comment]
I was thinking a blog post in the Arrow project ;)

**Comment by BryanCutler** [comment]
retest this please

File: R/pkg/R/SQLContext.R
Comment (HyukjinKwon) [inline R/pkg/R/SQLContext.R]: Assuming from ARROW-3204, Arrow is still not in the CRAN and looks it's going to take few months, and looks we can run the build via AppVeyor when it's on CRAN.

**Comment by HyukjinKwon** [comment]
@felixcheung and @shivaram, are you okay with this plan https://github.com/apache/spark/pull/22954#issuecomment-457214769 ? If so, I think we can go ahead.

File: R/pkg/R/SQLContext.R
Comment (felixcheung) [reply R/pkg/R/SQLContext.R]: yes, published to CRAN not happening yet for Arrow. install_github should be fine? it's a one time thing

**Comment by felixcheung** [comment]
Yes, install_github is fine. I’d say #2 gapply etc is higher priority. Sounds good to me.

**Comment by HyukjinKwon** [comment]
Yea will do. Do you mind if we go ahead with this PR @felixcheung?

**Comment by felixcheung** [comment]
Sure.

**Comment by HyukjinKwon** [comment]
Thanks. @felixcheung.

Merged to master.

**Comment by HyukjinKwon** [comment]
BTW, https://issues.apache.org/jira/browse/SPARK-26759 has subtasks for Arrow optimization (just FYI if anyone missed it)

**Pull request closed without merging**
